src/JSON/Deserializer.dfy

```diff
@@ -39,68 +39,91 @@ module {:options "-functionSyntax:4"} JSON.Deserializer {
     js.At(0) == 't' as byte
   }
 
+  function UnsupportedEscape16(code: seq<uint16>): DeserializationError {
+    UnsupportedEscape(FromUTF16Checked(code).UnwrapOr("Couldn't decode UTF-16"))
+  }
+
+  module Uint16StrConversion refines Str.ParametricConversion {
+    import opened BoundedInts
+
+    type Char = uint16
+  }
+
+  const HEX_TABLE_16: map<Uint16StrConversion.Char, nat> :=
+    map[
+      '0' as uint16 := 0, '1' as uint16 := 1, '2' as uint16 := 2, '3' as uint16 := 3, '4' as uint16 := 4,
+      '5' as uint16 := 5, '6' as uint16 := 6, '7' as uint16 := 7, '8' as uint16 := 8, '9' as uint16 := 9,
+      'a' as uint16 := 0xA, 'b' as uint16 := 0xB, 'c' as uint16 := 0xC, 'd' as uint16 := 0xD, 'e' as uint16 := 0xE, 'f' as uint16 := 0xF,
+      'A' as uint16 := 0xA, 'B' as uint16 := 0xB, 'C' as uint16 := 0xC, 'D' as uint16 := 0xD, 'E' as uint16 := 0xE, 'F' as uint16 := 0xF
+    ]
+
+  function ToNat16(str: Uint16StrConversion.String): uint16
+    requires |str| <= 4
+    requires forall c | c in str :: c in HEX_TABLE_16
+  {
+    Uint16StrConversion.ToNat_bound(str, 16, HEX_TABLE_16);
+    var hd := Uint16StrConversion.ToNat_any(str, 16, HEX_TABLE_16);
+    assert hd < 0x1_0000 by { reveal Pow(); }
+    hd as uint16
+  }
+
   // TODO: Verify this function
-  function Unescape(str: string, start: nat := 0): DeserializationResult<string>
+  function {:tailrecursion} {:vcs_split_on_every_assert} Unescape(str: seq<uint16>, start: nat := 0, prefix: seq<uint16> := []): DeserializationResult<seq<uint16>>
```

Comment:
Done, not because I don't think it would be valuable to add more specification to this function, but because that will be quite a bit of work. I also don't want to imply there's no verification happening in this code.